src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/CollectionExpressionExtensions.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions;
+
+internal static class CollectionExpressionExtensions
+{
+    public static bool IsWellKnownCollectionInterface(ITypeSymbol type)
+        => IsWellKnownCollectionReadOnlyInterface(type) || IsWellKnownCollectionReadWriteInterface(type);
+
+    public static bool IsWellKnownCollectionReadOnlyInterface(ITypeSymbol type)
+    {
+        return type.OriginalDefinition.SpecialType
+            is SpecialType.System_Collections_Generic_IEnumerable_T
+            or SpecialType.System_Collections_Generic_IReadOnlyCollection_T
+            or SpecialType.System_Collections_Generic_IReadOnlyList_T;
+    }
+
+    public static bool IsWellKnownCollectionReadWriteInterface(ITypeSymbol type)
+    {
+        return type.OriginalDefinition.SpecialType
+            is SpecialType.System_Collections_Generic_ICollection_T
+            or SpecialType.System_Collections_Generic_IList_T;
+    }
+
+    public static bool IsConstructibleCollectionType(
+        Compilation compilation,
+        [NotNullWhen(true)] ITypeSymbol? type)
+    {
+        return IsConstructibleCollectionType(compilation, type, out _);
+    }
+
+    public static bool IsConstructibleCollectionType(
+        Compilation compilation,
+        [NotNullWhen(true)] ITypeSymbol? type,
+        out ImmutableArray<ITypeSymbol> elementTypes)
```

Comment:
It might be worth documenting why this is a collection versus a single element type -- I guess because for a dictionary you have key/value?

Reply:
actually, this should be a singl etype.  let me fix that.